Harden linux/windows vault implementation for concurrent access.

aleksmaus opened the pull request:
## What does this PR do?

Harden linux/windows vault implementation for concurrent access.

#### Problem details

During repetitive E2E tests runs (took over 5 hours to reproduce) the file based vault implementation (that is used for linux and Windows) got into the state where the key value was corrupted, the resulted in the error:
```
Enrollment failed after failed to store agent config: could not save enrollment information: failed to ensure key: 
could not get agent key: could not decrypt key: cipher: message authentication failed
```
It's difficult to reproduce in the normal agent usage scenario, but looking at the implementation there is no mitigation against multiple processes accessing vault. The lower level API allows to reproduce the situation artificially see ```TestVaultConcurrent``` test https://github.com/elastic/elastic-agent/compare/main...aleksmaus:feature/harden_fsvault?expand=1#diff-8a5388ab5e868a37ee07711c7978dc51561a302be8fa560946038e71375fe70bR183

#### Solution

Harden linux/windows vault implementation for concurrent access.

* Use "atomic" write/rename for the data files https://github.com/elastic/elastic-agent/compare/main...aleksmaus:feature/harden_fsvault?expand=1#diff-5176984040d0726ff96ca6e27e372355b57829a8f657c706fd6960a6e4f96a45R17
* Use file locking https://github.com/elastic/elastic-agent/compare/main...aleksmaus:feature/harden_fsvault?expand=1#diff-6c9bf532380b28932cd23bc8d308c07263fa77a517d41fc0e2fd390d5e361a9eR71. The ```.lock``` file is created inside of the ```vault``` directory now.
* For the file locking API needed to pass context.Context down to the storage, that lead to the storage API change and the code that uses the storage is modified in order to pass the context down.
* Removed ```migrate_secret.go``` that doesn't seem to be used anywhere anymore.

## Why is it important?

Hardens the vault data agains corruption when it's being accessed concurrently. 

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)
- [x] I have added an integration test or an E2E test

elasticmachine commented:
## :grey_exclamation: Build Aborted
  
  > There is a new build on-going so the previous on-going builds have been aborted.
  


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/elastic-agent%2Felastic-agent-mbp%2FPR-3240/detail/PR-3240/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/elastic-agent%2Felastic-agent-mbp%2FPR-3240/detail/PR-3240/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/elastic-agent%2Felastic-agent-mbp%2FPR-3240/detail/PR-3240/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://fleet-ci.elastic.co/blue/organizations/jenkins/elastic-agent%2Felastic-agent-mbp%2FPR-3240/detail/PR-3240/1//artifacts)   [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/fleet-ci/transactions/view?rangeFrom=2023-08-15T02:06:04.996Z&rangeTo=2023-08-15T02:26:04.996Z&transactionName=BUILD+elastic-agent%2Felastic-agent-mbp%2FPR-%7Bnumber%7D&transactionType=job&latencyAggregationType=avg&traceId=5d3ae59e6f34892da6f079994339c117&transactionId=34d8589fcf409bff)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2023-08-15T02:16:04.996+0000


* Duration: 12 min 0 sec




#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 1  |
| Passed       | 0  |
| Skipped      | 0 |
| Total        | 1   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->

  ### Test errors [![1](https://img.shields.io/badge/1%20-red)](https://fleet-ci.elastic.co/blue/organizations/jenkins/elastic-agent%2Felastic-agent-mbp%2FPR-3240/detail/PR-3240/1//tests)
<details><summary>Expand to view the tests failures</summary><p>
    
    
##### `[empty] – TEST-go-unit.xml`


<ul>
<li>no error details</li>
<details><summary>Expand to view the stacktrace</summary><p>

```
 Test report file /var/lib/jenkins/workspace/_agent_elastic-agent-mbp_PR-3240/src/github.com/elastic/elastic-agent/build/TEST-go-unit.xml was length 0 
 ```
</p></details>
</ul>
    
</p></details>


<!-- STEPS ERRORS IF ANY -->
















## :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.


- `/package` : Generate the packages.


- `run integration tests` : Run the Elastic Agent Integration tests.


- `run end-to-end tests` : Generate the packages and run the E2E Tests.


- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)




</p>
</details>

<!--COMMENT_GENERATED_WITH_ID_comment.id-->

elasticmachine commented:
### :globe_with_meridians: Coverage report

Name | Metrics % (`covered/total`) | Diff
 --- | --- | ---
Packages | 98.734% (`78/79`) | :+1: 
Files | 67.018% (`191/285`) | :-1: -0.236
Classes | 65.472% (`347/530`) | :-1: -0.124
Methods | 52.431% (`1089/2077`) | :-1: -0.048
Lines | 38.261% (`12420/32461`) | :-1: -0.107
Conditionals | 100.0% (`0/0`) | :green_heart: 

<!--COMMENT_GENERATED_WITH_ID_coverage-->

elasticmachine commented:
Pinging @elastic/elastic-agent (Team:Elastic-Agent)

aleksmaus commented on internal/pkg/agent/migration/migrate_secret.go:
```diff
@@ -1,163 +0,0 @@
-	darwin = "darwin"
-)
-
-// MigrateAgentSecret migrates agent secret if the secret doesn't exists agent upgrade from 8.3.0 - 8.3.2 to 8.x and above on Linux and Windows platforms.
```
I have no idea, but it's not called from anywhere.

cmacknz replied:
Looks like we lost this in the big 8.6 merge https://github.com/elastic/elastic-agent/blob/006784462a006cc31fbba70ec43522500c1816bb/internal/pkg/agent/cmd/run.go#L132-L140

Here was the original bug: https://github.com/elastic/elastic-agent/issues/682

Would need to confirm if that situation is still possible with the change in directory structure in 8.6 and above.

aleksmaus replied:
Whooops. Been doing ok without it for few releases. 🤷

aleksmaus commented on internal/pkg/agent/vault/vault_linux.go:
```diff
@@ -157,3 +224,40 @@ func deriveKey(pw []byte, salt []byte) ([]byte, []byte, error) {
+
+// writeFile "atomic" file write, utilizes temp file and replace
+// os.CreateTemp creates the file with 0600 mask, which is what we need
+func writeFile(fp string, data []byte) (err error) {
```
the code is fairly small, will it keep the vault package separate from other agent libs. makes it easier to extract, put into different smaller project to test cross-platform etc.

AndersonQ commented on internal/pkg/agent/vault/vault_linux.go:
```diff
@@ -59,42 +65,90 @@ func New(path string, opts ...OptionFunc) (v *Vault, err error) {
+	if err != nil {
+		return fmt.Errorf("failed to acquire exclusive lock: %v, err: %w", v.lock.Path(), err)
+	}
+	return err
```
[Blocker]
Please, `err` is always `nil` so return `nil`.
```suggestion
	return nil
```

aleksmaus replied:
done

AndersonQ commented on internal/pkg/agent/vault/vault_linux.go:
```diff
@@ -59,42 +65,90 @@ func New(path string, opts ...OptionFunc) (v *Vault, err error) {
+	if err != nil {
+		return fmt.Errorf("failed to acquire shared lock: %v, err: %w", v.lock.Path(), err)
+	}
+	return err
```
[Blocker]
Please, `err` is always `nil` so return `nil`.
```suggestion
	return nil
```

AndersonQ commented on internal/pkg/agent/vault/vault_linux.go:
```diff
@@ -59,42 +65,90 @@ func New(path string, opts ...OptionFunc) (v *Vault, err error) {
 
-// Close closes the valut store
+// unlock unlocks the file lock and preserves the original error if there was a error
+func (v *Vault) unlock(err error) error {
+	unerr := v.lock.Unlock()
+	if err != nil {
+		return err
+	}
+	return unerr
+}
```
[Blocker]
Both errors are relevant. Better to use `errors.Join` and return both of them.
You can even just pass the errors to `errors.Join` and it'll take care of omitting the `nil` errors and returning `nil` if all errors are `nil`

```suggestion
func (v *Vault) unlock(err error) error {
	unerr := v.lock.Unlock()
	if err != nil {
		return err
	}
	return errors.Join(unerr, err)
}
```

AndersonQ commented on internal/pkg/agent/vault/vault_linux.go:
```diff
@@ -157,3 +224,40 @@ func deriveKey(pw []byte, salt []byte) ([]byte, []byte, error) {
 	return filepath.Join(v.path, fileNameFromKey(v.key, key))
 }
+
+// writeFile "atomic" file write, utilizes temp file and replace
+// os.CreateTemp creates the file with 0600 mask, which is what we need
+func writeFile(fp string, data []byte) (err error) {
+	dir, fn := filepath.Split(fp)
+	if dir == "" {
+		dir = "."
+	}
+
+	f, err := os.CreateTemp(dir, fn)
+	if err != nil {
+		return fmt.Errorf("failed creating temp file: %w", err)
+	}
+	defer func() {
```
[Suggestion]
1 - move the comment about the `os.CreateTem` to where it's used.
2 - I think the sentence could be improved

```suggestion
// writeFile performs an "atomic" file write operation, using a temporary file
// and then replacing the original.
func writeFile(fp string, data []byte) (err error) {
	dir, fn := filepath.Split(fp)
	if dir == "" {
		dir = "."
	}

	// os.CreateTemp creates the file with a 0600 permission mask,
	// which is the required permission setting.
	f, err := os.CreateTemp(dir, fn)
	if err != nil {
		return fmt.Errorf("failed creating temp file: %w", err)
	}
	defer func() {
```

AndersonQ commented on internal/pkg/agent/vault/vault_linux.go:
```diff
@@ -157,3 +224,40 @@ func deriveKey(pw []byte, salt []byte) ([]byte, []byte, error) {
+	}
+	defer func() {
+		if err != nil {
+			_ = os.Remove(f.Name())
```
[Blocker]
Please add this to the error as well.
You can join this error with the other and an extra message to clearly states this error is from the cleanup.

```suggestion
			rerr := os.Remove(f.Name())
			if rerr !=nil{
				err= errors.Join(err, fmt.Errorf("cleanup failed, could not remove temp file: %w", rerr))
			}
```

AndersonQ commented on internal/pkg/agent/vault/vault_options_nondarwin.go:
```diff
@@ -0,0 +1,23 @@
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+//go:build linux || windows
```
If it's _nondarwin_, why not use:
```suggestion
//go:build !darwin
```

AndersonQ commented on internal/pkg/agent/vault/vault_test.go:
```diff
@@ -164,3 +174,73 @@ func TestVault(t *testing.T) {
+		return fmt.Errorf("failed to set secret: %w", err)
+	}
+
+	sec, err := v.Get(ctx, key)
+	if err != nil {
+		return fmt.Errorf("failed to get secret: %w", err)
+	}
+
+	_ = sec
```
[Suggestion]
```suggestion
	_, err = v.Get(ctx, key)
	if err != nil {
		return fmt.Errorf("failed to get secret: %w", err)
	}
```

AndersonQ commented on internal/pkg/agent/vault/vault_test.go:
```diff
@@ -164,3 +174,73 @@ func TestVault(t *testing.T) {
+	}
+}
+
+func doCrud(t *testing.T, ctx context.Context, vaultPath, key string) error {
```
[Suggestion]
I don't see `t` being used, so you could remove it.
```suggestion
func doCrud(ctx context.Context, vaultPath, key string) error {
```

AndersonQ commented on internal/pkg/agent/vault/vault_windows.go:
```diff
@@ -59,42 +65,90 @@ func New(path string, opts ...OptionFunc) (v *Vault, err error) {
 	}
 
-	entropy, err := getOrCreateSeed(path, options.readonly)
+	r := &Vault{
```
[NIT]
Why are you using `r` instead of `v`?

aleksmaus replied:
want to return v as nil if err happens after this line

AndersonQ commented on internal/pkg/agent/vault/vault_windows.go:
```diff
@@ -59,42 +65,90 @@ func New(path string, opts ...OptionFunc) (v *Vault, err error) {
+
+// unlock unlocks the file lock and preserves the original error if there was a error
+func (v *Vault) unlock(err error) error {
+	unerr := v.lock.Unlock()
+	if err != nil {
+		return err
+	}
+	return unerr
```
[Blocker]
Same as in the other one. Please return both errors, you can use `errors.Join` for that.

AndersonQ commented on internal/pkg/agent/vault/vault_windows.go:
I'm wondering if it'd be worth having a common file for those vault methods that seems to be the same for windows and linux and a _darwin file with the darwin specific implementation

aleksmaus replied:
done

pierrehilbert commented:
Buildkite is failing on main because we are unable to deploy 8.11.0 on Elastic Cloud for now. @AndersonQ is working on a workaround to unlock us in the meantime.

AndersonQ commented on internal/pkg/agent/vault/vault_notdarwin.go:
```diff
@@ -0,0 +1,214 @@
+		return nil, err
+	}
+	defer func() {
+		err = r.unlock(err)
```
[Blocker]
It'll override the error of `getOrCreateSeed`. Better to `errors.Join` the errors.

aleksmaus replied:
it will not, it's a helper func using errors.Join

AndersonQ commented on internal/pkg/agent/vault/vault_notdarwin.go:
```diff
@@ -0,0 +1,214 @@
+		return nil, err
+	}
+	defer func() {
+		err = v.unlock(err)
```
[Blocker]
Same here, it'll override any other error

AndersonQ commented on internal/pkg/agent/vault/vault_notdarwin.go:
```diff
@@ -0,0 +1,214 @@
+	return nil
+}
+
+// unlock unlocks the file lock and preserves the original error if there was a error
+func (v *Vault) unlock(err error) error {
+	return errors.Join(err, v.lock.Unlock())
+}
```
[Blocker]
I don't understand why an unlock function needs to receive an error. The caller is responsible to handle any previous error and then decide what to do if `unlock` fails as well.

Besides it harms the single responsibility principle, unlock is also responsible to aggregate an error for no obvious reason.

aleksmaus replied:
do not agree with making it a blocker, it's a helper function to make the deferred call code above more compact and consistent, since it's done in multiple places

AndersonQ commented on internal/pkg/agent/vault/vault_notdarwin.go:
```diff
@@ -0,0 +1,214 @@
+		err = v.unlock(err)
+	}()
+
+	if _, err = os.Stat(v.filepathFromKey(key)); err == nil {
+		ok = true
+	} else if errors.Is(err, fs.ErrNotExist) {
+		err = nil
+	}
+	return ok, err
```
[Suggestion]
Avoid `if else` for a clearer, more explicit code
 
```suggestion
	if _, err = os.Stat(v.filepathFromKey(key)); err != nil {
		if errors.Is(err, fs.ErrNotExist) {
			return false, nil
		}
		return false, err
	}
	
	return true, nil
```

AndersonQ commented on internal/pkg/agent/vault/vault_notdarwin.go:
```diff
@@ -0,0 +1,214 @@
+		return err
+	}
+	defer func() {
+		err = v.unlock(err)
```
[Blocker]
same as the others

aleksmaus replied:
explained above why this is done the way it's done

AndersonQ commented on internal/pkg/agent/vault/vault_notdarwin.go:
```diff
@@ -0,0 +1,214 @@
+	return options
+}
+
+// fileNameFromKey returns the filename as a hash of the vault seed combined with the key
+// this ties the key with the vault seed eliminating the change of attempting
+// to decrypt the key for the wrong vault seed value.
```
```suggestion
// fileNameFromKey returns the filename as a hash of the vault seed combined with the key.
// This ties the key with the vault seed eliminating the chance of attempting
// to decrypt the key for the wrong vault seed value.
```

AndersonQ commented on internal/pkg/agent/vault/vault_windows.go:
```diff
@@ -152,3 +43,7 @@ func systemAdministratorsOnly(path string, inherit bool) error {
 		acl.GrantSid(0xF10F0000, administratorsSID))
 }
+
+func writeFile(fp string, data []byte) (err error) {
+	return os.WriteFile(fp, data, 0600)
+}
```
[Suggestion]
```suggestion
func writeFile(fp string, data []byte) error {
	return os.WriteFile(fp, data, 0600)
}
```

aleksmaus commented:
> just one small blocker, I marked it as `[Blocker]`

addressed the comments. do not agree with the blocker, added explanation why, it's a private helper function to avoid repeating the same code in 5 places. could you review again?

cmacknz commented:
This is going to go into 8.10 first and then get forward ported to main. This will get us around the issue with the 8.11 snapshot.

AndersonQ commented:
Merged with failing integration tests because this PR fixes another integration tests issue affecting all tests installing the agent. Therefore we need to have it merged to reduce the number of failing tests and focus on one issue at a time.
